src/operations/blas1/WGAtomicReduction.hpp

```diff
@@ -0,0 +1,137 @@
+WGAtomicReduction<operator_t, lhs_t, rhs_t>::WGAtomicReduction(lhs_t& _l,
+                                                               rhs_t& _r)
+    : lhs_(_l), rhs_(_r){};
+
+template <typename operator_t, typename lhs_t, typename rhs_t>
+PORTBLAS_INLINE typename WGAtomicReduction<operator_t, lhs_t, rhs_t>::index_t
+WGAtomicReduction<operator_t, lhs_t, rhs_t>::get_size() const {
+  return rhs_.get_size();
+}
+
+template <typename operator_t, typename lhs_t, typename rhs_t>
+PORTBLAS_INLINE bool WGAtomicReduction<operator_t, lhs_t, rhs_t>::valid_thread(
+    cl::sycl::nd_item<1> ndItem) const {
+  return true;
+}
+
+template <typename operator_t, typename lhs_t, typename rhs_t>
+PORTBLAS_INLINE typename WGAtomicReduction<operator_t, lhs_t, rhs_t>::value_t
+WGAtomicReduction<operator_t, lhs_t, rhs_t>::eval(cl::sycl::nd_item<1> ndItem) {
+  auto atomic_res = sycl::atomic_ref<value_t, sycl::memory_order::relaxed,
+                                     sycl::memory_scope::device,
+                                     sycl::access::address_space::global_space>(
+      lhs_.get_data()[0]);
+  const auto size = get_size();
+  int lid = ndItem.get_global_linear_id();
+  value_t val = operator_t::template init<rhs_t>();
+  const auto loop_stride =
+      ndItem.get_local_range(0) * ndItem.get_group_range(0);
+
+  // First loop for big arrays
+  for (int id = lid; id < size; id += loop_stride) {
+    val = operator_t::eval(val, rhs_.eval(id));
+  }
+
+  val = sycl::reduce_over_group(ndItem.get_sub_group(), val, sycl::plus<>());
+
+  if ((ndItem.get_local_id() &
+       (ndItem.get_sub_group().get_local_range() - 1)) == 0) {
+    atomic_res += val;
+  }
+  return {};
+}
+template <typename operator_t, typename lhs_t, typename rhs_t>
+template <typename sharedT>
+PORTBLAS_INLINE typename WGAtomicReduction<operator_t, lhs_t, rhs_t>::value_t
+WGAtomicReduction<operator_t, lhs_t, rhs_t>::eval(sharedT scratch,
+                                                  cl::sycl::nd_item<1> ndItem) {
+  auto atomic_res = sycl::atomic_ref<value_t, sycl::memory_order::relaxed,
+                                     sycl::memory_scope::device,
+                                     sycl::access::address_space::global_space>(
+      lhs_.get_data()[0]);
+  const auto size = get_size();
+  const int lid = static_cast<int>(ndItem.get_global_linear_id());
+  const auto loop_stride =
+      ndItem.get_local_range(0) * ndItem.get_group_range(0);
+  value_t val = operator_t::template init<rhs_t>();
+
+  // First loop for big arrays
+  for (int id = lid; id < size; id += loop_stride) {
+    val = operator_t::eval(val, rhs_.eval(id));
+  }
+
+  val = sycl::reduce_over_group(ndItem.get_sub_group(), val, sycl::plus<>());
+
+  if (ndItem.get_sub_group().get_local_id() == 0) {
+    scratch[ndItem.get_sub_group().get_group_linear_id()] = val;
+  }
+  ndItem.barrier();
+
+  val = (ndItem.get_local_id() < (ndItem.get_local_range(0) /
```

Comment:
Also, is this getting used for IAmin/IAmax. For that, I think you cannot just put 0, you need to use whatever initial `value_t` for that would be the init for tuple Index.

Reply:
No, this is only for `asum` at the moment. We plan to iteratively apply this new kernel to other operators later on.